packages/core/src/shared/utilities/map.ts

```diff
@@ -0,0 +1,104 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from '../logger'
+import type { KeyedCache } from './cacheUtils'
+
+/**
+ * Simply maps a Key to an Object. Everything is done in memory, allowing for atomic updates.
+ *
+ * There is a common use case for this class, so look to reuse this when possible.
+ * Benefits, instead of making your own:
+ * - A key does not have to be a string, {@link NestedMap.asKey} enables this
```

Comment:
(followup) this is called "hash" usually, isn't it? we can look for ways to give clues to other developers, by using common terms. 

```suggestion
 * - A key does not have to be a string, {@link NestedMap.hash} enables this
```